compiler/src/dmd/expression.d

```diff
@@ -6996,6 +6997,31 @@ extern (C++) final class PrettyFuncInitExp : DefaultInitExp
     }
 }
 
+/***********************************************************
+ * Named argument in an argument list
+ *
+ * Examples: `f(x: 3)`, `g!(T: int)`
+ */
+extern (C++) class NamedArgExp : Expression
```

Comment:
It might make the semantic analysis easier if (I just though of this now) if a `NamedArgExp` is actually a call expression with named arguments.

Reply:
Currently this being part of the expression hierarchy sounds like a bad idea

Reply:
The argument matching needs to be done lexically *and* semantically so having them as expressions has some potential for wacky shit to happen

Reply:
I don't know what you mean with "argument matching needs to be done lexically *and* semantically".

Reply:
> It might make the semantic analysis easier if (I just though of this now) if a `NamedArgExp` is actually a call expression with named arguments.

A `CallExp` has an `Expressions* arguments`, where would I store the names of the named arguments if not in a `NamedArgExp`?

Reply:
In a struct explicitly for managing the pairs of names and expressions.

Reply:
Normal arguments are matched only semantically, then passed onto the backend after the correct overload has been found, named arguments have to matched lexically *then* semantically *then* lowered to a traditional argument list (which is then evaluated in the lexical order of the original expressions).

Reply:
If you do all of this as expressions you will have to overwrite them (thus making error messages impossible) or hack some godawful mess into the AST

Reply:
Most of the semantic pass is about rewriting the parse time AST into an AST for code gen, overwriting `NamedArgExp` is not unusual. I don't know what error you have in mind that becomes impossible or messy.

Reply:
> In a struct explicitly for managing the pairs of names and expressions.

You don't think named arguments are and `ASTNode`? Not even for dmd as a library?

Reply:
Trust me it'll get absolutely horrible if you just do it all in a array of "expressions" rather than something more targeted.

Reply:
They probably should be an ASTNode since Parameter (i.e. for the visitor pattern potentially) but ideally they'd only be accessed via a fairly carefully laid out API

Reply:
Implementing the semantic analysis for function calls to normal functions is fairly easy, implementing the semantic analysis to anything with an overload (ESPECIALLY a template overload) is absolutely hell on earth because it tries to save references to the arguments all over the place so if you start overwriting them piecemeal as you go through them stuff just randomly breaks.

Reply:
And think about it, you overwrite the array, any error message, AST dump etc., now prints something which is complete gibberish since it's printing what the compiler lowered the arguments into not how the user wrote them down.

Reply:
`NamedArgExp` as proposed here is nice, for example this can be reused for struct literals or I dont know what else in the future (theoritically).

Reply:
Still not an expression

Reply:
Indeed, you have to follow the DIP.